core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/columns/Utils.kt

```diff
@@ -146,25 +167,31 @@ internal fun <C> ColumnsContainer<*>.getColumn(path: ColumnPath, policy: Unresol
             UnresolvedColumnsPolicy.Create -> DataColumn.empty().cast<C>()
         }
 
-internal fun <T> List<ColumnWithPath<T>>.top(): List<ColumnWithPath<T>> {
-    val root = TreeNode.createRoot<ColumnWithPath<T>?>(null)
-    forEach { root.put(it.path, it) }
-    return root.topDfs { it.data != null }.map { it.data!! }
+/**
+ * Returns a sub-list of columns that are roots of the trees of columns.
+ *
+ * In practice, this means that if a column in [this] is a child of another column in [this],
+ * it will not be included in the result.
+ */
+internal fun <T> List<ColumnWithPath<T>>.roots(): List<ColumnWithPath<T>> {
+    val emptyRoot = TreeNode.createRoot<ColumnWithPath<T>?>(data = null)
+    this.forEach { emptyRoot.put(it.path, it) }
+    return emptyRoot.topmostChildren { it.data != null }.map { it.data!! }
 }
 
 internal fun List<ColumnWithPath<*>>.allColumnsExcept(columns: Iterable<ColumnWithPath<*>>): List<ColumnWithPath<*>> {
     if (isEmpty()) return emptyList()
     val fullTree = collectTree()
     columns.forEach {
         var node = fullTree.getOrPut(it.path).asNullable()
-        node?.dfs()?.forEach { it.data = null }
+        node?.allChildren()?.forEach { it.data = null }
         while (node != null) {
             node.data = null
             node = node.parent
         }
     }
-    val dfs = fullTree.topDfs { it.data != null }
-    return dfs.map { it.data!!.addPath(it.pathFromRoot()) }
+    val subtrees = fullTree.topmostChildren { it.data != null }
```

Comment:
what would you suggest?

Reply:
As my doc explains it takes the tree and returns first the children, closest to the root of the tree, where the predicate holds. It doesn't matter whether the children of those nodes also hold, those are ignored. So since we now have `allChildren { something == true }`, I thought `topmostChildren { something == true }`. It can also be viewed as subtree-roots where the predicate holds.